File: 2026-03/src/day_09/boj_14956/Main.cpp
#include <bits/stdc++.h>
using namespace std;

pair<int, int> dfs(int n, int m) {
    if (n == 1) return {1, 1};

    int len = n / 2;
    int area = len * len;

    int q = (m - 1) / area;
    int nxt = (m - 1) % area + 1;

    auto [x, y] = dfs(n / 2, nxt);

    if (q == 0) return {y, x};
    if (q == 1) return {x, y + len};
    if (q == 2) return {x + len, y + len};
    return {2 * len - y + 1, len - x + 1};
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, m;
    cin >> n >> m;

    auto [x, y] = dfs(n, m);
    cout << x << ' ' << y;
}

File: 2026-03/src/day_09/boj_14956/Main.java
package day_09.boj_14956;

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[] pos = dfs(n, m);
        System.out.println(pos[0] + " " + pos[1]);
    }

    static int[] dfs(int n, int m) {
        if (n == 1) return new int[]{1, 1};

        int len = n / 2;
        int area = len * len;

        int q = (m - 1) / area;
        int nxt = (m - 1) % area + 1;

        int[] p = dfs(n / 2, nxt);
        int x = p[0];
        int y = p[1];

        if (q == 0) return new int[]{y, x};
        if (q == 1) return new int[]{x, y + len};
        if (q == 2) return new int[]{x + len, y + len};
        return new int[]{2 * len - y + 1, len - x + 1};
    }
}
